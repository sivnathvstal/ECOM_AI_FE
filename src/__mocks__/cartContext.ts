import { Cart, CartProduct } from '../types';

export const mockCartItems: CartProduct[] = [
  {
    productId: 1,
    quantity: 2,
    product: {
      id: 1,
      title: 'Test Product 1',
      price: 10.99,
      description: 'Test description 1',
      category: 'electronics',
      image: 'test1.jpg',
      rating: { rate: 4.5, count: 100 },
    },
  },
  {
    productId: 2,
    quantity: 1,
    product: {
      id: 2,
      title: 'Test Product 2',
      price: 20.99,
      description: 'Test description 2',
      category: 'clothing',
      image: 'test2.jpg',
      rating: { rate: 4.0, count: 80 },
    },
  },
];

export const mockCart: Cart = {
  id: 1,
  userId: 1,
  date: '2023-01-01',
  products: [
    { productId: 1, quantity: 2 },
    { productId: 2, quantity: 1 },
  ],
};

export const mockCartContext = {
  cart: mockCart,
  cartItems: mockCartItems,
  isLoading: false,
  error: null,
  addItem: jest.fn(),
  updateItem: jest.fn(),
  removeItem: jest.fn(),
  totalItems: 3, // 2 + 1
  totalPrice: 42.97, // 10.99*2 + 20.99*1
};

export const mockEmptyCartContext = {
  cart: {
    id: 0,
    userId: 1,
    date: '2023-01-01',
    products: [],
  },
  cartItems: [],
  isLoading: false,
  error: null,
  addItem: jest.fn(),
  updateItem: jest.fn(),
  removeItem: jest.fn(),
  totalItems: 0,
  totalPrice: 0,
};

export const mockLoadingCartContext = {
  ...mockCartContext,
  isLoading: true,
};

export const mockErrorCartContext = {
  ...mockCartContext,
  error: 'Failed to fetch cart',
};

const defaultExport = {
  mockCartItems,
  mockCart,
  mockCartContext,
  mockEmptyCartContext,
  mockLoadingCartContext,
  mockErrorCartContext,
};
export default defaultExport; 