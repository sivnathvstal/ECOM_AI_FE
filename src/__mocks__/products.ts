import { Product, PaginationParams, ProductsResponse } from '../types';

export const mockProducts: Product[] = [
  {
    id: 1,
    title: 'Test Product 1',
    price: 10.99,
    description: 'Test description 1',
    category: 'electronics',
    image: 'test1.jpg',
    rating: { rate: 4.5, count: 100 },
  },
  {
    id: 2,
    title: 'Test Product 2',
    price: 20.99,
    description: 'Test description 2',
    category: 'clothing',
    image: 'test2.jpg',
    rating: { rate: 4.0, count: 80 },
  },
  {
    id: 3,
    title: 'Test Product 3',
    price: 15.99,
    description: 'Test description 3',
    category: 'electronics',
    image: 'test3.jpg',
    rating: { rate: 3.5, count: 60 },
  },
  {
    id: 4,
    title: 'Test Product 4',
    price: 25.99,
    description: 'Test description 4',
    category: 'jewelry',
    image: 'test4.jpg',
    rating: { rate: 4.8, count: 120 },
  },
];

export const mockPagination: PaginationParams = {
  page: 1,
  limit: 8,
  total: 4,
};

export const mockProductsResponse: ProductsResponse = {
  products: mockProducts,
  pagination: mockPagination,
};

export const mockCategories: string[] = [
  'electronics',
  'clothing',
  'jewelry',
]; 