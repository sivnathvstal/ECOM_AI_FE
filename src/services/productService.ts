import api from './api';
import { Product, ProductsResponse } from '../types';

export const getProducts = async (
  page = 1,
  limit = 8,
  category?: string
): Promise<ProductsResponse> => {
  try {
    // First get all products to calculate total
    const allProductsResponse = await api.get<Product[]>('/products');
    const allProducts = allProductsResponse.data;
    
    // Filter by category if provided
    const filteredProducts = category 
      ? allProducts.filter((product) => product.category === category) 
      : allProducts;
    
    const total = filteredProducts.length;
    
    // Calculate pagination
    const startIndex = (page - 1) * limit;
    const endIndex = startIndex + limit;
    const paginatedProducts = filteredProducts.slice(startIndex, endIndex);
    
    return {
      products: paginatedProducts,
      pagination: {
        page,
        limit,
        total
      }
    };
  } catch (error) {
    console.error('Error fetching products:', error);
    throw error;
  }
};

export const getProductById = async (id: number): Promise<Product> => {
  try {
    const response = await api.get<Product>(`/products/${id}`);
    return response.data;
  } catch (error) {
    console.error(`Error fetching product with id ${id}:`, error);
    throw error;
  }
};

export const getCategories = async (): Promise<string[]> => {
  try {
    const response = await api.get<string[]>('/products/categories');
    return response.data;
  } catch (error) {
    console.error('Error fetching categories:', error);
    throw error;
  }
};

export const searchProducts = async (
  query: string,
  page = 1,
  limit = 8
): Promise<ProductsResponse> => {
  try {
    const allProductsResponse = await api.get<Product[]>('/products');
    const allProducts = allProductsResponse.data;
    
    // Filter products by search query
    const filteredProducts = allProducts.filter((product) => 
      product.title.toLowerCase().includes(query.toLowerCase()) ||
      product.description.toLowerCase().includes(query.toLowerCase())
    );
    
    const total = filteredProducts.length;
    
    // Calculate pagination
    const startIndex = (page - 1) * limit;
    const endIndex = startIndex + limit;
    const paginatedProducts = filteredProducts.slice(startIndex, endIndex);
    
    return {
      products: paginatedProducts,
      pagination: {
        page,
        limit,
        total
      }
    };
  } catch (error) {
    console.error('Error searching products:', error);
    throw error;
  }
}; 