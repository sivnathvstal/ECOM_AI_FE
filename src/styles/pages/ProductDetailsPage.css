.product-details-container {
  padding: 16px;
  margin-top: 16px;
}

.product-details-paper {
  padding: 24px;
}

.product-image {
  width: 100%;
  max-height: 400px;
  object-fit: contain;
}

.product-details {
  display: flex;
  flex-direction: column;
  height: 100%;
}

.product-title {
  margin-bottom: 16px;
}

.product-price {
  color: #3f51b5; /* primary color */
  font-weight: bold;
  font-size: 1.5rem;
  margin-bottom: 16px;
}

.product-description {
  margin-bottom: 24px;
  flex-grow: 1;
}

.add-to-cart-button {
  margin-top: auto;
}

.loader-container {
  display: flex;
  justify-content: center;
  padding: 32px;
}

.back-button {
  margin-bottom: 16px;
} 