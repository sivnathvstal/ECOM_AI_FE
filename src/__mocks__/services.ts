import { mockProducts, mockProductsResponse, mockCategories } from './products';
import { mockCart, mockCartItems } from './cartContext';

// Product service mocks
export const mockProductService = {
  getProducts: jest.fn().mockResolvedValue(mockProductsResponse),
  getProductById: jest.fn().mockImplementation((id: number) => {
    const product = mockProducts.find(p => p.id === id);
    if (product) {
      return Promise.resolve(product);
    }
    return Promise.reject(new Error('Product not found'));
  }),
  getCategories: jest.fn().mockResolvedValue(mockCategories),
  searchProducts: jest.fn().mockResolvedValue(mockProductsResponse),
};

// Cart service mocks
export const mockCartService = {
  getCart: jest.fn().mockResolvedValue(mockCart),
  addToCart: jest.fn().mockResolvedValue(mockCart),
  updateCartItem: jest.fn().mockResolvedValue(mockCart),
  removeFromCart: jest.fn().mockResolvedValue(mockCart),
  getCartWithProducts: jest.fn().mockResolvedValue(mockCartItems),
}; 