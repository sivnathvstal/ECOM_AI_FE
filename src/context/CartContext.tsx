import React, { createContext, useState, useContext, useEffect, ReactNode } from 'react';
import { Cart, CartProduct } from '../types';
import { getCart, addToCart, updateCartItem, removeFromCart, getCartWithProducts } from '../services/cartService';

interface CartContextType {
  cart: Cart;
  cartItems: CartProduct[];
  isLoading: boolean;
  error: string | null;
  addItem: (productId: number, quantity: number) => Promise<void>;
  updateItem: (productId: number, quantity: number) => Promise<void>;
  removeItem: (productId: number) => Promise<void>;
  totalItems: number;
  totalPrice: number;
}

const CartContext = createContext<CartContextType | undefined>(undefined);

export const useCart = (): CartContextType => {
  const context = useContext(CartContext);
  if (!context) {
    throw new Error('useCart must be used within a CartProvider');
  }
  return context;
};

interface CartProviderProps {
  children: ReactNode;
}

export const CartProvider: React.FC<CartProviderProps> = ({ children }) => {
  const [cart, setCart] = useState<Cart>({
    id: 0,
    userId: 1,
    date: new Date().toISOString(),
    products: []
  });
  const [cartItems, setCartItems] = useState<CartProduct[]>([]);
  const [isLoading, setIsLoading] = useState<boolean>(true);
  const [error, setError] = useState<string | null>(null);

  const fetchCart = async () => {
    try {
      setIsLoading(true);
      const cartData = await getCart();
      setCart(cartData);
      const cartWithProducts = await getCartWithProducts();
      setCartItems(cartWithProducts);
      setError(null);
    } catch (err) {
      setError('Failed to fetch cart');
      console.error(err);
    } finally {
      setIsLoading(false);
    }
  };

  useEffect(() => {
    fetchCart();
  }, []);

  const addItem = async (productId: number, quantity: number) => {
    try {
      setIsLoading(true);
      const updatedCart = await addToCart(productId, quantity);
      setCart(updatedCart);
      await fetchCart(); // Refetch to get product details
    } catch (err) {
      setError('Failed to add item to cart');
      console.error(err);
    } finally {
      setIsLoading(false);
    }
  };

  const updateItem = async (productId: number, quantity: number) => {
    try {
      setIsLoading(true);
      const updatedCart = await updateCartItem(productId, quantity);
      setCart(updatedCart);
      await fetchCart(); // Refetch to get product details
    } catch (err) {
      setError('Failed to update cart item');
      console.error(err);
    } finally {
      setIsLoading(false);
    }
  };

  const removeItem = async (productId: number) => {
    try {
      setIsLoading(true);
      const updatedCart = await removeFromCart(productId);
      setCart(updatedCart);
      await fetchCart(); // Refetch to get product details
    } catch (err) {
      setError('Failed to remove item from cart');
      console.error(err);
    } finally {
      setIsLoading(false);
    }
  };

  // Calculate total items in cart
  const totalItems = cartItems.reduce((total, item) => total + item.quantity, 0);

  // Calculate total price
  const totalPrice = cartItems.reduce((total, item) => {
    return total + ((item.product?.price || 0) * item.quantity);
  }, 0);

  const value = {
    cart,
    cartItems,
    isLoading,
    error,
    addItem,
    updateItem,
    removeItem,
    totalItems,
    totalPrice
  };

  return <CartContext.Provider value={value}>{children}</CartContext.Provider>;
}; 