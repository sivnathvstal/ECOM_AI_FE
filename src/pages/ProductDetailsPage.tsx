import React, { useState, useEffect } from 'react';
import { useParams } from 'react-router-dom';
import {
  Container,
  Grid,
  Typography,
  Button,
  Box,
  CircularProgress,
  Paper
} from '@material-ui/core';
import { ShoppingCart } from '@material-ui/icons';
import { getProductById } from '../services/productService';
import { useCart } from '../context/CartContext';
import { Product } from '../types';
import './ProductDetailsPage.css';

interface ParamTypes {
  id: string;
}

const ProductDetailsPage: React.FC = () => {
  const { id } = useParams<ParamTypes>();
  const [product, setProduct] = useState<Product | null>(null);
  const [loading, setLoading] = useState<boolean>(true);
  const [error, setError] = useState<string | null>(null);
  const { addItem, isLoading: cartLoading } = useCart();

  useEffect(() => {
    const fetchProduct = async () => {
      try {
        setLoading(true);
        const data = await getProductById(parseInt(id, 10));
        setProduct(data);
        setError(null);
      } catch (err) {
        setError('Failed to fetch product details');
        console.error(err);
      } finally {
        setLoading(false);
      }
    };

    fetchProduct();
  }, [id]);

  const handleAddToCart = async () => {
    if (product) {
      await addItem(product.id, 1);
    }
  };

  if (loading) {
    return (
      <Container className="product-details-container" data-testid="product-details-loading">
        <Box className="loader-container">
          <CircularProgress />
        </Box>
      </Container>
    );
  }

  if (error || !product) {
    return (
      <Container className="product-details-container" data-testid="product-details-error">
        <Typography color="error">
          {error || 'Product not found'}
        </Typography>
      </Container>
    );
  }

  return (
    <Container className="product-details-container" data-testid="product-details-page">
      <Paper className="product-details-paper">
        <Grid container spacing={4}>
          <Grid item xs={12} md={6}>
            <img 
              src={product.image} 
              alt={product.title} 
              className="product-image"
              data-testid="product-image"
            />
          </Grid>
          <Grid item xs={12} md={6}>
            <Box className="product-details">
              <Typography 
                variant="h4" 
                component="h1" 
                className="product-title"
                data-testid="product-title"
              >
                {product.title}
              </Typography>
              <Typography 
                variant="h5" 
                component="p" 
                className="product-price"
                data-testid="product-price"
              >
                ${product.price.toFixed(2)}
              </Typography>
              <Typography 
                variant="body1" 
                component="div" 
                className="product-description"
                data-testid="product-description"
              >
                {product.description}
              </Typography>
              <Button
                variant="contained"
                color="primary"
                size="large"
                startIcon={<ShoppingCart />}
                onClick={handleAddToCart}
                disabled={cartLoading}
                className="add-to-cart-button"
                data-testid="add-to-cart-button"
              >
                Add to Cart
              </Button>
            </Box>
          </Grid>
        </Grid>
      </Paper>
    </Container>
  );
};

export default ProductDetailsPage; 