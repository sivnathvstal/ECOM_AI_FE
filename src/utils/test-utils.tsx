import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { ThemeProvider, createTheme } from '@material-ui/core';
import { QueryClient, QueryClientProvider } from 'react-query';
import { CartProvider } from '../context/CartContext';

// Create a theme for testing
const theme = createTheme({
  palette: {
    primary: {
      main: '#3f51b5',
    },
    secondary: {
      main: '#f50057',
    },
  },
});

// Create a client for React Query
const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
});

// All providers wrapper
const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <QueryClientProvider client={queryClient}>
      <ThemeProvider theme={theme}>
        <CartProvider>
          <BrowserRouter>{children}</BrowserRouter>
        </CartProvider>
      </ThemeProvider>
    </QueryClientProvider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

// Re-export everything
export * from '@testing-library/react';

// Override render method
export { customRender as render }; 