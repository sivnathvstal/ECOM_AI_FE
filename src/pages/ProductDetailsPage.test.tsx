import React from 'react';
import { render, screen, fireEvent, waitFor, act } from '../utils/test-utils';
import { Route, MemoryRouter } from 'react-router-dom';
import ProductDetailsPage from './ProductDetailsPage';
import * as productService from '../services/productService';
import * as cartContext from '../context/CartContext';
import { mockProducts } from '../__mocks__/products';
import { mockCartContext } from '../__mocks__/cartContext';

// Mock the productService
jest.mock('../services/productService', () => ({
  getProductById: jest.fn(),
}));

// Mock the useCart hook
jest.mock('../context/CartContext', () => ({
    ...jest.requireActual('../context/CartContext'),
  useCart: jest.fn(),
}));

describe('ProductDetailsPage', () => {
  const product = mockProducts[0];
  
  beforeEach(() => {
    jest.clearAllMocks();
    (productService.getProductById as jest.Mock).mockResolvedValue(product);
    (cartContext.useCart as jest.Mock).mockReturnValue(mockCartContext);
  });

  const renderWithRouter = async (id: number) => {
    let container;
    await act(async () => {
      const rendered = render(
        <MemoryRouter initialEntries={[`/product/${id}`]}>
          <Route path="/product/:id">
            <ProductDetailsPage />
          </Route>
        </MemoryRouter>
      );
      container = rendered.container;
    });
    return { container };
  };

  it('renders correctly', async () => {
    const { container } = await renderWithRouter(product.id);
    await waitFor(() => expect(productService.getProductById).toHaveBeenCalled());
    expect(container).toMatchSnapshot();
  });

  it('displays loading state initially', async () => {
    renderWithRouter(product.id);
    expect(screen.getByTestId('product-details-loading')).toBeInTheDocument();
  });

  it('fetches and displays product details', async () => {
    await renderWithRouter(product.id);
    
    // Wait for product to be fetched
    await waitFor(() => expect(productService.getProductById).toHaveBeenCalledWith(product.id));
    
    // Check if product details are rendered
    expect(screen.getByTestId('product-title')).toHaveTextContent(product.title);
    expect(screen.getByTestId('product-price')).toHaveTextContent(`$${product.price.toFixed(2)}`);
    expect(screen.getByTestId('product-description')).toHaveTextContent(product.description);
    
    const image = screen.getByTestId('product-image');
    expect(image).toHaveAttribute('src', product.image);
    expect(image).toHaveAttribute('alt', product.title);
  });

  it('displays error message when API call fails', async () => {
    const error = new Error('Failed to fetch product details');
    (productService.getProductById as jest.Mock).mockRejectedValue(error);
    
    await renderWithRouter(999); // Non-existent product ID
    
    await waitFor(() => {
      expect(screen.getByTestId('product-details-error')).toBeInTheDocument();
      expect(screen.getByText('Failed to fetch product details')).toBeInTheDocument();
    });
  });

  it('displays "Product not found" when product is null', async () => {
    (productService.getProductById as jest.Mock).mockResolvedValue(null);
    
    await renderWithRouter(999); // Non-existent product ID
    
    await waitFor(() => {
      expect(screen.getByTestId('product-details-error')).toBeInTheDocument();
      expect(screen.getByText('Product not found')).toBeInTheDocument();
    });
  });

  it('adds product to cart when "Add to Cart" button is clicked', async () => {
    await renderWithRouter(product.id);
    
    // Wait for product to be fetched
    await waitFor(() => expect(productService.getProductById).toHaveBeenCalled());
    
    // Click the "Add to Cart" button
    await act(async () => {
      fireEvent.click(screen.getByTestId('add-to-cart-button'));
    });
    
    // Check if addItem was called with the correct parameters
    expect(mockCartContext.addItem).toHaveBeenCalledWith(product.id, 1);
  });

  it('disables "Add to Cart" button when cart is loading', async () => {
    (cartContext.useCart as jest.Mock).mockReturnValue({
      ...mockCartContext,
      isLoading: true
    });
    
    await renderWithRouter(product.id);
    
    // Wait for product to be fetched
    await waitFor(() => expect(productService.getProductById).toHaveBeenCalled());
    
    // Check if the "Add to Cart" button is disabled
    const addToCartButton = screen.getByTestId('add-to-cart-button');
    expect(addToCartButton).toBeDisabled();
  });
}); 