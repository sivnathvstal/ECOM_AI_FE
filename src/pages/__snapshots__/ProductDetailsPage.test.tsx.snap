// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetailsPage renders correctly 1`] = `
<div>
  <div
    class="MuiContainer-root product-details-container MuiContainer-maxWidthLg"
    data-testid="product-details-page"
  >
    <div
      class="MuiPaper-root product-details-paper MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
        >
          <img
            alt="Test Product 1"
            class="product-image"
            data-testid="product-image"
            src="test1.jpg"
          />
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
        >
          <div
            class="MuiBox-root MuiBox-root-2 product-details"
          >
            <h1
              class="MuiTypography-root product-title MuiTypography-h4"
              data-testid="product-title"
            >
              Test Product 1
            </h1>
            <p
              class="MuiTypography-root product-price MuiTypography-h5"
              data-testid="product-price"
            >
              $
              10.99
            </p>
            <div
              class="MuiTypography-root product-description MuiTypography-body1"
              data-testid="product-description"
            >
              Test description 1
            </div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained add-to-cart-button MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge"
              data-testid="add-to-cart-button"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                <span
                  class="MuiButton-startIcon MuiButton-iconSizeLarge"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zM1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2H1zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
                    />
                  </svg>
                </span>
                Add to Cart
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
